tools/amd_build/pyHIPIFY/hipify-python.py

```diff
@@ -0,0 +1,967 @@
+
+                    # Overwrite file contents
+                    fileobj.seek(0)
+                    fileobj.write(new_output_source)
+                    fileobj.truncate()
+                    fileobj.flush()
+
+                    # Flush to disk
+                    os.fsync(fileobj)
+
+
+def main():
+    """Example invocation
+
+    python hipify.py --project-directory /home/myproject/ --extensions cu cuh h cpp --output-directory /home/gains/
+    """
+
+    parser = argparse.ArgumentParser(
+        description="The Python Hipify Script.")
+
+    parser.add_argument(
+        '--project-directory',
+        type=str,
+        default=os.getcwd(),
+        help="The root of the project.",
+        required=True)
+
+    parser.add_argument(
+        '--show-detailed',
+        type=bool,
+        default=False,
+        help="Show detailed summary of the hipification process.",
+        required=False)
+
+    parser.add_argument(
+        '--extensions',
+        nargs='+',
+        default=["cu", "cuh", "c", "cpp", "h", "in", "hpp"],
+        help="The extensions for files to run the Hipify script over.",
+        required=False)
+
+    parser.add_argument(
+        '--output-directory',
+        type=str,
+        default="",
+        help="The directory to store the hipified project.",
+        required=False)
+
+    parser.add_argument(
+        '--include-dirs',
+        nargs='+',
+        default=[],
+        help="The directories under the root that should be included.",
+        required=False)
+
+    parser.add_argument(
+        '--yaml-settings',
+        type=str,
+        default="",
+        help="The yaml file storing information for disabled functions and modules.",
+        required=False)
+
+    parser.add_argument(
+        '--add-static-casts',
+        type=bool,
+        default=False,
+        help="Whether to automatically add static_casts to kernel arguments.",
+        required=False)
+
+    args = parser.parse_args()
+
+    # Verify the project directory exists.
+    if not os.path.exists(args.project_directory):
+        print("The project folder specified does not exist.")
+        return
+
+    # Make sure output directory exists.
+    if not os.path.exists(args.output_directory):
+        print("The output folder already exists.")
+        return
+
+    # If no output directory, provide a default one.
+    if args.output_directory is "":
+        args.project_directory = args.project_directory[0:-
+                                                        1] if args.project_directory.endswith("/") else args.project_directory
+        args.output_directory = args.project_directory + "_amd"
+
+    # Copy from project directory to output directory if not done already.
+    if not os.path.exists(args.output_directory):
+        shutil.copytree(args.project_directory, args.output_directory)
+
+    # Extract all of the kernel parameter and template type information.
```

Comment:
This comment seems inconsistent with the conditional that comes below it.

Reply:
Perhaps it wanted to also add, "if necessary"? Why does this have to be done now, as opposed to later?

Reply:
I think the comment does reflect what the code is doing: if we enable the "add static_cast" feature, the script is extracting template and kernel parameters. Do you disagree?

Reply:
concerning the if necessary: the current logic will add static_casts for all arguments, even the ones that match the type definition (script is easier that way). Concerning the now and not later: it'll first build the kernel arguments up and then does the static_cast addition in a second pass.